src/server_state.cpp

```diff
@@ -1080,11 +1128,26 @@ void wsrep::server_state::wait_until_state(
     while (state_ != state)
     {
         cond_.wait(lock);
+        // If the waiter waits for any other state than disconnecting
+        // or disconnected and the state has been changed to disconnecting,
+        // this usually means that some error was encountered 
+        if (state != s_disconnecting && state != s_disconnected
```

Comment:
What happens if we change provider variable when something is waiting for other states (non s_disconnecting or s_disconnected)?